### PR TITLE
Allow configuration of depth of headings for search

### LINKED ISSUE #270: Include level 4 headings in search results
Current: headings of level 1-3 are auto-included in search results.

Suggestion: allow level 4 headings in search results

Justification: While this can increase search results, I noticed that some of the important keywords e.g, `focus groups`, `prototyping` in se-edu/se-book are in level 4 headings. It is not possible to raise them to level 3.

On a related note, may be we can allow the depth of headings auto-included in search results to be specified in site.json?

### PR DESCRIPTION
**Pull request by nusjzx**
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Documentation update
• [ ] Bug fix
• [ ] New feature
• [X] Enhancement to an existing feature
• [ ] Other, please explain:

Resolves #270

**What is the rationale for this request?**
User should customize the depth of headings based on their website content

**What changes did you make? (Give an overview)**
accept `depthOfHeadings` in site.json, include until `depthOfHeadings` in the site data

**Testing instructions:**
1.`markbind init`
2.insert following content in index.md:
```
<searchbar :data="searchData" placeholder="Search" :on-hit="searchCallback"></searchbar>
<h1 id ="h11">h11</h1>
<h2 id ="h21">h21</h2>
<h1 id ="h12">h12</h1>
<h2 id ="h22">h22</h2>
<h3 id ="h31">h31</h3>
<h3 id ="h32">h32</h3>
<h1 id ="h13">h13</h1>
<h2 id ="h23">h23</h2>
<h3 id ="h33">h33</h3>
<h3 id ="h34">h34</h3>
<h4 id ="h14">h14</h4>
<h4 id ="h24">h24</h4>
```
3. customize depthOfHeadings in site.json and test search result.

### CODE REVIEW

File: docs/userGuide/siteConfiguration.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ Let's examine a typical `site.json` file:
 | **pages** | An array of pages to be rendered. The `src` is the file; `title` is the page title for the generated web page. Titles specified here take priority over titles specified in the [front matter](contentAuthoring.html#front-matter) of individual pages. Alternatively, `glob` can be used to define a file pattern. If `title` is not specified in the front matter, the page will have `titlePrefix` as its title. |
 | **ignore** | An array of file patterns to be ignored. By default, MarkBind will copy all the files as assets into the output folder. The ignore pattern follows the [glob pattern used in .gitignore](https://git-scm.com/docs/gitignore#_pattern_format). For example, `*.md` ignores all markdown source files. You may want to ignore the Git directory `.git/*`. |
 | **deploy** | The settings for [auto deployment to Github pages](ghpagesDeployment.html). The `message` is the commit message used for the deployment commit. |
+| **depthOfHeadings** | A number that specifies depth of headings for searching
```
Comment (yamgent) [inline docs/userGuide/siteConfiguration.md]: Suggested change: `The depth of headings to be included for searching. Default: <code>3</code>. | `

* Try to be more consistent with the style of the rest of the properties.
* There is a default level of headings, so we should mention it.
* Missing fullstop and `|`.

File: lib/Page.js [hunk 2]
```diff
@@ -60,6 +60,7 @@ function Page(pageConfig) {
   this.frontMatter = {};
   this.headings = {};
   this.includedFiles = {};
+  this.depthOfHeadings = pageConfig.depthOfHeadings
```
Comment (yamgent) [inline lib/Page.js]: Missing semi-colon at the end. Do check the coding style by running `npm run lintwin` frequently. Travis is failing as well due to this issue.

File: lib/Page.js [hunk 3]
```diff
@@ -168,7 +169,11 @@ Page.prototype.prepareTemplateData = function () {
  */
 Page.prototype.collectHeadings = function (renderedPage) {
   const $ = cheerio.load(renderedPage);
-  $('h1,h2,h3').each((i, heading) => {
+  let headingsSelector = '';
+  for (let i = 1; i <= this.depthOfHeadings; i += 1) {
+    headingsSelector += `h${i}, `;
+  }
+  $(headingsSelector.slice(0, -2)).each((i, heading) => {
```
Comment (yamgent) [inline lib/Page.js]: We can avoid the slicing by doing the following:
```javascript
...

if (this.depthOfHeadings > 0) {
  let headingsSelector = 'h1';
  for (let i = 2; i <= this.depthOfHeadings; i += 1) {
    headingsSelector += `, h${i}`;
  }
  $(headingsSelector).each((i, heading) => {
}

...
```

File: lib/Site.js [hunk 4]
```diff
@@ -255,6 +255,7 @@ Site.prototype.createPage = function (config) {
     src: config.pageSrc,
     title: config.title || '',
     titlePrefix: this.siteConfig.titlePrefix,
+    depthOfHeadings: this.siteConfig.depthOfHeadings || 3,
```
Comment (yamgent) [inline lib/Site.js]: Let's not use magic numbers. Specify a constant at the top of the file. Maybe `DEPTH_OF_HEADINGS_DEFAULT`?

**Comment by yamgent** [comment]
Also update your PR title:

`Allow configuration of depth of headings for search`

* Should not contain the number issue.
* Try to use full words whenever possible.

File: lib/Site.js [hunk 5]
```diff
@@ -25,6 +25,7 @@ const Page = require('./Page');
 const CLI_VERSION = require('../package.json').version;
 
 const BOILERPLATE_FOLDER_NAME = '_markbind/boilerplates';
+const DEPTH_OF_HEADINGS_DEFAULT = 3
```
Comment (yamgent) [inline lib/Site.js]: Missing semicolon

File: lib/Page.js [hunk 6]
```diff
@@ -168,9 +169,16 @@ Page.prototype.prepareTemplateData = function () {
...
-  $('h1,h2,h3').each((i, heading) => {
-    this.headings[$(heading).attr('id')] = $(heading).text();
-  });
+  let headingsSelector = '';
+  if (this.depthOfHeadings > 0) {
+    let headingsSelector = 'h1';
+    for (let i = 1; i <= this.depthOfHeadings; i += 1) {
+      headingsSelector += `, h${i}`;
+    }
+    $(headingsSelector.slice(0, -2)).each((i, heading) => {
```
Comment (yamgent) [inline lib/Page.js]: As mentioned, slice is no longer needed, as the algorithm has changed to no longer include trailing commas.

File: lib/Page.js [hunk 7]
```diff
@@ -168,9 +169,16 @@ Page.prototype.prepareTemplateData = function () {
  */
 Page.prototype.collectHeadings = function (renderedPage) {
   const $ = cheerio.load(renderedPage);
-  $('h1,h2,h3').each((i, heading) => {
-    this.headings[$(heading).attr('id')] = $(heading).text();
-  });
+  let headingsSelector = '';
```
Comment (yamgent) [inline lib/Page.js]: This variable declaration is not necessary.

File: lib/Page.js [hunk 8]
```diff
@@ -168,9 +169,16 @@ Page.prototype.prepareTemplateData = function () {
...
+  if (this.depthOfHeadings > 0) {
+    let headingsSelector = 'h1';
+    for (let i = 1; i <= this.depthOfHeadings; i += 1) {
```
Comment (yamgent) [inline lib/Page.js]: Should start with 2, not 1 (`h1` is already in `headingsSelector`)

File: docs/userGuide/siteConfiguration.md [hunk 9]
```diff
@@ -48,6 +48,7 @@ Let's examine a typical `site.json` file:
 | **pages** | An array of pages to be rendered. The `src` is the file; `title` is the page title for the generated web page. Titles specified here take priority over titles specified in the [front matter](contentAuthoring.html#front-matter) of individual pages. Alternatively, `glob` can be used to define a file pattern. If `title` is not specified in the front matter, the page will have `titlePrefix` as its title. |
 | **ignore** | An array of file patterns to be ignored. By default, MarkBind will copy all the files as assets into the output folder. The ignore pattern follows the [glob pattern used in .gitignore](https://git-scm.com/docs/gitignore#_pattern_format). For example, `*.md` ignores all markdown source files. You may want to ignore the Git directory `.git/*`. |
 | **deploy** | The settings for [auto deployment to Github pages](ghpagesDeployment.html). The `message` is the commit message used for the deployment commit. |
+| **depthOfHeadings** | The depth of headings to be included for searching. Default: `3`. |
```
Comment (yamgent) [inline docs/userGuide/siteConfiguration.md]: You might want to include an entry in the example code above.

Comment (damithc) [reply docs/userGuide/siteConfiguration.md]: `depthOfHeadings` is not self explanatory enough. `headingIndexingLevel`?

On a related note, we may want to support this at page level too i.e., in a glob/src entry. But that can be a future PR.

File: docs/userGuide/siteConfiguration.md [hunk 10]
```diff
@@ -36,7 +36,8 @@ Let's examine a typical `site.json` file:
   ],
   "deploy": {
     "message": "Site Update."
-  }
+  },
+  "headingIndexingLevel": 4
```
Comment (yamgent) [inline docs/userGuide/siteConfiguration.md]: Use the default value: 3.

File: docs/userGuide/siteConfiguration.md [hunk 11]
```diff
@@ -48,6 +49,7 @@ Let's examine a typical `site.json` file:
 | **pages** | An array of pages to be rendered. The `src` is the file; `title` is the page title for the generated web page. Titles specified here take priority over titles specified in the [front matter](contentAuthoring.html#front-matter) of individual pages. Alternatively, `glob` can be used to define a file pattern. If `title` is not specified in the front matter, the page will have `titlePrefix` as its title. |
 | **ignore** | An array of file patterns to be ignored. By default, MarkBind will copy all the files as assets into the output folder. The ignore pattern follows the [glob pattern used in .gitignore](https://git-scm.com/docs/gitignore#_pattern_format). For example, `*.md` ignores all markdown source files. You may want to ignore the Git directory `.git/*`. |
 | **deploy** | The settings for [auto deployment to Github pages](ghpagesDeployment.html). The `message` is the commit message used for the deployment commit. |
+| **headingIndexingLevel** | The index level of headings to be included for searching. Default: `3`. |
```
Comment (yamgent) [inline docs/userGuide/siteConfiguration.md]: Sounds a bit weird to say "index level of headings". Maybe something like:

```
The level of headings to be indexed for searching.
```

File: lib/Site.js [hunk 12]
```diff
@@ -25,6 +25,7 @@ const Page = require('./Page');
 const CLI_VERSION = require('../package.json').version;
 
 const BOILERPLATE_FOLDER_NAME = '_markbind/boilerplates';
+const HEADING_INDEXING_LEVEL_DEFAULT = 3;
```
Comment (yamgent) [inline lib/Site.js]: 1. Let's update the test to check that the default heading indexing level is adhered to.
1. We can also modify the inner site to test for custom indexing level (4?).

File: docs/userGuide/siteConfiguration.md (hunk 10 above)
Comment (acjh) [reply docs/userGuide/siteConfiguration.md]: @yamgent Setting this to `3` is redundant. It's sufficient to mention the default in the table below.

**Pull request merged**